core/virtualhostname/virtualhostname.go

```diff
@@ -74,6 +130,28 @@ func (i Info) Target() HostnameTarget {
 	return i.target
 }
 
+// String returns the virtual hostname stringified.
+func (i Info) String() string {
+	switch i.target {
+	case MachineTarget:
+		return fmt.Sprintf("%d.%s.%s", i.machine, i.modelUUID, Domain)
+	case UnitTarget:
+		parts := strings.Split(i.unit, "/")
+		if len(parts) != 2 {
+			panic("invalid unit name") // this shouldn't happen because we have validated the unit in the constructor.
```

Comment:
todo: Replace the panic. Even if this should never happen, do not panic the juju controller on purpose. Replace at line 147 also.